### ISSUE TITLE
Add support for better (but attached to costs) sources

### ISSUE THREAD

**Issue by greentux** [ev-map/EVMap#306]
The 2 available sourcen goingelectric / open chargemap are free but with a lot of problems (user maintained). In the future this will become more and more inaccurate. So perhaps an option for an charged source would be helpful.

**Comment by johan12345** [comment]
Hmm, I understand what you mean, but:

**a) what data source would actually be better?** 
Official data from the charging station operators can be obtained through the roaming networks, for Europe this would most likely be [Hubject](https://www.hubject.com/). Hubject's pricing seems to be prohibitively expensive though (especially the 4-figure setup fees), even if this were a paid feature in the app. The data quality is also not perfect, e.g. often the listed charging power is inaccurate, or a whole charging site can be listed sometimes as a group and sometimes as a separate entry for each single charger. 

The [Google Places API](https://cloud.google.com/blog/products/maps-platform/introducing-the-new-places-api-with-access-to-new-ev-accessibility-features-and-more?hl=en) also provides EV charger locations now, but from what I saw in the Google Maps app, data quality is not very good (some duplicates, often inaccurate operator information, many chargers without realtime status). Pricing here is per request, so there's no fixed upfront or monthly payment. But still, with a large amount of users, this would not be insignificant.

And then we have other companies that also build charging apps themselves, like [Cirrantic](https://cirrantic.com/) (operators of the [Moovility](https://moovility.me/) app, data is probably mostly from Hubject and partially directly from operators) or [PlugShare](https://company.plugshare.com/api.html) (community-maintained data, just like GoingElectric and OCM). I don't know about their pricing.

**b) what benefit would it bring to EVMap?**
Data from the roaming networks usually just gives you very basic information like location, operator, available plugs and power, EVSEIDs and possibly realtime status. All this information is already available in any of the charging providers' apps. So what benefit would it have to show exactly the same data in EVMap, especially if the users would need to pay for something that they get for free in other apps?

I think one of the main benefits of EVMap is exactly the more detailed information (like photos, descriptions, fault reports, compatibility with different charging cards) that can only be contributed by users and is rarely available in the charging providers' apps. If we want to keep this, we would either have to aggregate free and commercial data sources together (which is not trivial to do well [^1]) or build a new community-maintained data source ourselves (which could be subject to the same problems as the existing ones, maybe even more due to the smaller user base).

**c) What does it mean for the open source project?**
The more paid APIs we use, the harder it becomes for people to contribute to the development of the app[^2]. There are already [not a lot](https://github.com/ev-map/EVMap/graphs/contributors) of external contributions (mostly translations) - but if a paid API were the primary source for charging stations, it would become pretty much impossible for others to contribute to the app.

Also, shouldn't we as an open source project rather try to promote contributions to community-maintained and open[^3] data sources? As the number of EV drivers is growing, if we can convince enough of them to contribute, the data should not become more and more inaccurate.

[^1]: Especially with the considerations needed to make sure the GoingElectric API's terms of service are respected - as far as I understand it, the aggregation would need to be done locally in the app and not as a one-time job on a backend server. Also, data coming from different sources need to be clearly marked.
[^2]: We do already use some paid APIs (Chargeprice, fronyx), but both are not impacting the most basic functionality of the app, and at least Chargeprice provides a public staging API that can be used for testing.
[^3]: Unfortunately only OCM and OSM (which is currently being integrated in #290) are truly *open data* - GoingElectric's data is not released under an open license, and as mentioned, their API has pretty restrictive terms of use.

**Comment by greentux** [comment]
For me its the only charger mapping app an AAOS (maps is not on that level). But from time to time I recheck charger status on Moovility because of that charging station configuration hassle (distance between chargers not in sync with GE). i know its all community driven data, right.
But I am looking for a Moovility or mobilit+ AAOS app with additional data from GE :)

**Comment by johan12345** [comment]
Isn't mobility+ available on AAOS? They do have an Android Auto version at least...

Merging data from Moovility (or any other source mentioned above) with GE would run into the same problems as our current matching for the realtime status, because it would need to be done in basically the same way (mostly based on closest distance) as the GE API doesn't provide any data we can match unambiguously (such as the EVSEID). Of course the matching algorithm can still be improved (see #243), and it might be easier if we can process the data globally instead of one charger at a time (which, as I said, may not be allowed according to GE ToS), but it will probably never be perfect...

**Comment by 360path** [comment]
> Data from the roaming networks usually just gives you very basic information like location, operator, available plugs and power, EVSEIDs and possibly realtime status. All this information is already available in any of the charging providers' apps. So what benefit would it have to show exactly the same data in EVMap,

The big advantage of EVMap is that it aggregates chargers of multiple operators. Sure, some providers offer an app (also free) to find their chargers (and the ones in their partner network), but it's more than annoying to need to switch between apps. I prefer using EVMap, but not all chargers are there. I just recently became an EV driver, and there is a mismatch between the chargers shown through the app of the main provider of my home town (shows all of their chargers of course) GE (many of those), OCM (none of those). And I'm not sure whether I should add data to GE, or to OCM, or just use the provider's app because its data may be added through an API at some point (making all the work of adding the chargers to GE or OCM obsolete).

**Comment by johan12345** [comment]
> Sure, some providers offer an app (also free) to find their chargers (and the ones in their partner network), but it's more than annoying to need to switch between apps

Many of the big nationwide providers use Hubject as a roaming platform, so their partner network covers almost all chargers.
As I said, the main benefit of GE and OCM is that they often have more information than what the roaming networks can provide (photos, descriptions, etc.) and it is possible to add chargers that are not part of any roaming networks. But the downside is that the community has to maintain that information...

**Comment by greentux** [comment]
The best from two worlds. Charger data form Hubject and additional user content from GE / OCM.

**Comment by Linkinsoldier** [comment]
> Hmm, I understand what you mean, but:
> 
> **a) what data source would actually be better?** Official data from the charging station operators can be obtained through the roaming networks, for Europe this would most likely be [Hubject](https://www.hubject.com/). Hubject's pricing seems to be prohibitively expensive though (especially the 4-figure setup fees), even if this were a paid feature in the app. The data quality is also not perfect, e.g. often the listed charging power is inaccurate, or a whole charging site can be listed sometimes as a group and sometimes as a separate entry for each single charger.
> 
> The [Google Places API](https://cloud.google.com/blog/products/maps-platform/introducing-the-new-places-api-with-access-to-new-ev-accessibility-features-and-more?hl=en) also provides EV charger locations now, but from what I saw in the Google Maps app, data quality is not very good (some duplicates, often inaccurate operator information, many chargers without realtime status). Pricing here is per request, so there's no fixed upfront or monthly payment. But still, with a large amount of users, this would not be insignificant.
> 
> And then we have other companies that also build charging apps themselves, like [Cirrantic](https://cirrantic.com/) (operators of the [Moovility](https://moovility.me/) app, data is probably mostly from Hubject and partially directly from operators) or [PlugShare](https://company.plugshare.com/api.html) (community-maintained data, just like GoingElectric and OCM). I don't know about their pricing.
> 
> **b) what benefit would it bring to EVMap?** Data from the roaming networks usually just gives you very basic information like location, operator, available plugs and power, EVSEIDs and possibly realtime status. All this information is already available in any of the charging providers' apps. So what benefit would it have to show exactly the same data in EVMap, especially if the users would need to pay a monthly subscription for something that they get for free in other apps?
> 
> I think one of the main benefits of EVMap is exactly the more detailed information (like photos, descriptions, fault reports, compatibility with different charging cards) that can only be contributed by users and is rarely available in the charging providers' apps. If we want to keep this, we would either have to aggregate free and commercial data sources together (which is not trivial to do well [1](#user-content-fn-1-f9270df3e21e096f757a29aa2065cb3b)) or build a new community-maintained data source ourselves (which could be subject to the same problems as the existing ones, maybe even more due to the smaller user base).
> 
> **c) What does it mean for the open source project?** The more paid APIs we use, the harder it becomes for people to contribute to the development of the app[2](#user-content-fn-2-f9270df3e21e096f757a29aa2065cb3b). There are already [not a lot](https://github.com/ev-map/EVMap/graphs/contributors) of external contributions (mostly translations) - but if a paid API were the primary source for charging stations, it would become pretty much impossible for others to contribute to the app.
> 
> Also, shouldn't we as an open source project rather try to promote contributions to community-maintained and open[3](#user-content-fn-3-f9270df3e21e096f757a29aa2065cb3b) data sources? As the number of EV drivers is growing, if we can convince enough of them to contribute, the data should not become more and more inaccurate.
> ## Footnotes
> 
>     1. Especially with the considerations needed to make sure the GoingElectric API's terms of service are respected - as far as I understand it, the aggregation would need to be done locally in the app in real time, and not as a one-time job on a backend server. Also, data coming from different sources need to be clearly marked. [↩](#user-content-fnref-1-f9270df3e21e096f757a29aa2065cb3b)
> 
>     2. We do already use some paid APIs (Chargeprice, fronyx), but both are not impacting the most basic functionality of the app, and at least Chargeprice provides a public staging API that can be used for testing. [↩](#user-content-fnref-2-f9270df3e21e096f757a29aa2065cb3b)
> 
>     3. Unfortunately only OCM and OSM (which is currently being integrated in [Implement OpenStreetMap data source #290](https://github.com/ev-map/EVMap/pull/290)) are truly _open data_ - GoingElectric's data is not released under an open license, and as mentioned, their API has pretty restrictive terms of use. [↩](#user-content-fnref-3-f9270df3e21e096f757a29aa2065cb3b)

I totally agree with your assumptions. My question however would be: hasn't this changed in the last years with the new AFIR laws in Europe? Isn't it necessary for the CPO to send important data (including price info) and couldn't this be used to display/verify chargers and also show price info? I would love price info similar to charging price (App)

I only found this service here on a quick search, but I think you know what I mean:

https://mobilithek.info/blog/release-juni-2025

https://solidroam.com/ocpi
https://solidroam.com/blog/afir-compliance-cpo-guide

Also I read that ChargingTime is using price info from going electric to display this in there app. Since evmap also uses going electric as a data source: could evmap display charging prices as well? I would love this (similar to adhoc map)

**Comment by cremor** [comment]
@Linkinsoldier See https://github.com/ev-map/EVMap/issues/320#issuecomment-3720614126 about AFIR. (The Hocsy project that is mentioned there uses the AFIR data sources.)

And as I understand it, that is already in progress here: https://github.com/ev-map/backend

**Comment by Linkinsoldier** [comment]
Thank you for the hint. Sounds awesome 😎

**Comment by johan12345** [comment]
> Also I read that ChargingTime is using price info from going electric to display this in there app. Since evmap also uses going electric as a data source: could evmap display charging prices as well?

If ChargingTime would be doing that, they would violate GoingElectric's API terms of service, which prohibits accessing any data that is not available through the public API (which the pricing data is not). Also, in my experience, GoingElectric's pricing data is often outdated/inaccurate.

Yes, I am planning to use AFIR data (and already working on that [here](https://github.com/ev-map/backend). This should include pricing information as well, though only ad-hoc prices, no subscription or roaming offers. AFIR data in Germany is also not complete yet, quite a few operators are still [not providing data](https://hocsy.eu/de/auswertungen/cpo_afir_meldestatus.html) or are [missing pricing information](https://hocsy.eu/de/auswertungen/cpo_afir_fehlende_preise.html).

**Comment by Linkinsoldier** [comment]
Awesome, thank you!

**Comment by Linkinsoldier** [comment]
I have one more idea: are you aware of the German App Ladefuchs?
https://codeberg.org/Ladefuchs/ladefuchs-react-native

Its FOSS and gives detailed pricing info including subscription and roaming. Maybe that could help as well?

Edit: sorry, looking into the app in more detail it looks like it uses the Chargeprice API to get the prices...

**Comment by johan12345** [comment]
Yes, I know Ladefuchs and my impression so far was also that they were using Chargeprice data. Looking at the [backend code](https://codeberg.org/Ladefuchs/ladefuchs-api) this seems to be no longer the case since version 2.0, rather they seem to be using (at least partially?) the API from Eco-Movement. That is also a paid service though.
